nixos/modules/services/misc/matrix-synapse.nix

```diff
@@ -7,118 +7,108 @@ let
-        }'') v)}
-      }
-    '') cfg.servers)}
-    }
-  }
-redaction_retention_period: ${toString cfg.redaction_retention_period}
-app_service_config_files: ${builtins.toJSON cfg.app_service_config_files}
 
-${cfg.extraConfig}
-'';
+  # This is organized to match the sections in
+  # https://github.com/matrix-org/synapse/blob/develop/docs/sample_config.yaml
+  # Not everything in the config is configurable directly via Nix, so
+  # use extraConfig or settings to extend.
+  yamlConfig = {
+    # Server
+    inherit (cfg) server_name public_baseurl listeners;
+    inherit (cfg) bind_port unsecure_port bind_host; # deprecated, but keeping for backwards compatibility
+    pid_file = "/run/matrix-synapse.pid";
+
+    # Homeserver blocking
+    inherit (cfg) redaction_retention_period;
+
+    # TLS
+    inherit (cfg) tls_certificate_path tls_private_key_path no_tls;
+
+    # Federation
+
+    # Caching
+    inherit (cfg) event_cache_size;
+
+    # Database
+    database = {
+      name = cfg.database_type;
+      args = cfg.database_args;
+    };
+
+    # Logging
+    inherit (cfg) verbose;
+    log_config = logConfigFile;
+
+    # Ratelimiting
+    inherit (cfg) rc_message rc_federation;
+
+    # Media Store
+    inherit (cfg) max_upload_size max_image_pixels dynamic_thumbnails;
+    media_store_path = "${cfg.dataDir}/media";
+  } // optionalAttrs cfg.url_preview_enabled {
+    url_preview_enabled = true;
+    url_preview_ip_range_blacklist = cfg.url_preview_ip_range_blacklist;
+    url_preview_ip_range_whitelist = cfg.url_preview_ip_range_whitelist;
+    url_preview_url_blacklist = cfg.url_preview_url_blacklist;
+  } // {
+
+    # Captcha
+    inherit (cfg) recaptcha_private_key recaptcha_public_key enable_registration_captcha;
+    recaptcha_siteverify_api = "https://www.google.com/recaptcha/api/siteverify";
+
+    # TURN
+    inherit (cfg) turn_uris turn_shared_secret turn_user_lifetime;
+
+    # Registration
+    inherit (cfg) enable_registration registration_shared_secret bcrypt_rounds allow_guest_access;
+
+    account_threepid_delegates = filterAttrs (_: v: v != null) {
+      inherit (cfg.account_threepid_delegates) email msisdn;
+    };
+
+    # Account Validity
+
+    # Metrics
+    inherit (cfg) enable_metrics report_stats;
+
+    # API Configuration
+    inherit (cfg) macaroon_secret_key app_service_config_files room_prejoin_state;
+
+    # Signing Keys
+    inherit (cfg) key_refresh_interval;
+    signing_key_path = "${cfg.dataDir}/homeserver.signing.key";
+
+    perspectives = {
+      servers = mapAttrs
+        (name: value: {
+          verify_keys = mapAttrs (name: value: { key = value; }) value;
+        })
+        cfg.servers;
+    };
+
+    # Single sign-on integration
+    # Push
+    # Rooms
+    # Opentracing
+    # Workers
+  };
+
+  configFile = pkgs.writeText "homeserver.yaml" (
+    generators.toYAML { } (filterAttrs (_: v: v != null)
+      (fold recursiveUpdate { } [ yamlConfig cfg.settings ])));
```

Comment:
Instead of doing this, set the defaults in the `config` section by assigning to `services.matrix-synapse` directly with `mkDefault`'s. Check out https://nixos.org/manual/nixos/stable/index.html#ex-settings-nix-representable for an example.

Reply:
I don't think I understand what you mean. I don't think I am setting any defaults here... it's just collapsing all of the individual options that already exist on the module into a single attrset so that it can be converted to YAML.

Reply:
Oh right. Still should be done this way though. E.g. see the how the ZNC module provides support for legacy options: https://github.com/NixOS/nixpkgs/blob/7de613db995c79685c216524b8b6f73a2521dc46/nixos/modules/services/networking/znc/options.nix#L231-L261

The reason it should be done this way is because otherwise if other modules try to access `config.services.matrix-synapse.settings`, they won't have the full view of the set options, but rather only a partial one that doesn't include the ones from `yamlConfig` here. And in general `recursiveUpdate` should be avoided in modules, because the module system already supports merging by itself, and a much better version of it.